core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java

```diff
@@ -0,0 +1,335 @@
+     */
+    @Override
+    public void subscribe(Subscriber<? super AsyncResponseTransformer<ResponseT, ResponseT>> downstreamSubscriber) {
+        if (downstreamSubscriber == null) {
+            // §1.09 Subscribe throw NPE on null subscriber
+            throw new NullPointerException("downstreamSubscriber must not be null");
+        }
+        this.downstreamSubscriber = downstreamSubscriber;
+        downstreamSubscriber.onSubscribe(new DownstreamSubscription());
+        if (maxElements < 0) {
+            // §1.04 (optional) must signal on error when fails
+            // §1.09 may reject calls to subscribe if publisher is unable or unwilling to serve them
+            // Rejection must trigger on error after onSubscribe
+            downstreamSubscriber.onError(new IllegalArgumentException("Maximum element to request must be positive"));
+        }
+    }
+
+    /**
+     * The subscription implementation for the subscriber to this SplittingTransformer.
+     */
+    private class DownstreamSubscription implements Subscription {
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                // §3.09 request zero must signal IllegalArgumentException
+                downstreamSubscriber.onError(new IllegalArgumentException("Amount requested must be positive"));
+                return;
+            }
+            if (isBounded && totalEmitted.get() >= maxElements) {
+                terminate();
+                return;
+            }
+            if (!isCancelled.get()) {
+                long newDemand = outstandingDemand.updateAndGet(current -> {
+                    if (Long.MAX_VALUE - current < n) {
+                        return Long.MAX_VALUE;
+                    }
+                    return current + n;
+                });
+                log.trace(() -> String.format("new outstanding demand: %s", newDemand));
+                if (newDemand > 0) {
+                    emit();
+                }
+            }
+        }
+
+        @Override
+        public void cancel() {
+            // called by downstream subscriber to notify no more data is needed, we are done
+            if (isCancelled.compareAndSet(false, true)) {
+                log.trace(() -> "Cancelling splitting transformer");
+                publisherToUpstream.complete();
+                // §3.13 Cancel must make the publisher eventually drop all references to the subscriber
+                downstreamSubscriber = null;
+            }
+        }
+    }
+
+    private void terminate() {
+        if (terminated.compareAndSet(false, true)) {
+            publisherToUpstream.complete();
+            downstreamSubscriber.onComplete();
+            // §3.13 Cancel must make the publisher eventually drop all references to the subscriber
+            downstreamSubscriber = null;
+        }
+    }
+
+    private void emit() {
+        if (emitting.compareAndSet(false, true)) {
```

Comment:
We should double check the state before we set emitting to false because new items might have come in since we decided to release the flag.

See https://github.com/aws/aws-sdk-java-v2/blob/master/utils/src/main/java/software/amazon/awssdk/utils/async/SimplePublisher.java#L216-L235